### PR TITLE
Proposal to detects OpenBSD cookie protection for functions which

### PR DESCRIPTION
**Pull request by devnexen**
is added with .openbsd.randomdata elf section.

### CODE REVIEW

**Comment by XVilka** [comment]
There are some regressions it seems.

**Comment by devnexen** [comment]
Even if I remove my changes I still reproduce the issue locally.

File: libr/bin/p/bin_elf.c [hunk 1]
```diff
@@ -1034,6 +1034,29 @@ static bool has_canary(RBinFile *bf) {
 	return ret;
 }
 
+static int has_retguard(RBinFile *bf) {
+	int ret = -1;
+	RList* sections_list = sections (bf);
+	RListIter *iter;
+	RBinSection *section;
+	if (sections_list) {
```
Comment (radare) [inline libr/bin/p/bin_elf.c]: unnecessary if

File: libr/bin/p/bin_elf.c [hunk 2]
```diff
@@ -1034,6 +1034,29 @@ static bool has_canary(RBinFile *bf) {
...
+		r_list_foreach (sections_list, iter, section) {
+			# define R_BIN_RANDOMDATA_RETGUARD_SZ 48
+			if (!strcmp(section->name, ".openbsd.randomdata")) {
```
Comment (radare) [inline libr/bin/p/bin_elf.c]: add space before (

File: libr/bin/p/bin_elf.c [hunk 3]
```diff
@@ -1034,6 +1034,29 @@ static bool has_canary(RBinFile *bf) {
...
+				// The retguard cookie adds 8 per return function inst.
+				if (section->size >= R_BIN_RANDOMDATA_RETGUARD_SZ) {
+					ret = 1;
```
Comment (radare) [inline libr/bin/p/bin_elf.c]: ret = (section->size >= R_BIN_RANDOMDATA_RETGUARD_SZ)

File: libr/bin/p/bin_elf.c [hunk 4]
```diff
@@ -1034,6 +1034,29 @@ static bool has_canary(RBinFile *bf) {
 	return ret;
 }
 
+static int has_retguard(RBinFile *bf) {
```
Comment (radare) [inline libr/bin/p/bin_elf.c]: i would go for bool instead of int. -1 case is not really giving too much informtion in here

Comment (devnexen) [reply libr/bin/p/bin_elf.c]: Intentional here because it s relevant only for OpenBSD here thus rabin2 does not display this info for other oses.

Comment (devnexen) [reply libr/bin/p/bin_elf.c]: To explain further, openbsd/clang can compile with -fret-protector adding thus a retguard protection for each return function into .openbsd.randomdata section here we try to indicate only in the case of OpenBSD if this binary had been compiled with this flag or not.

Comment (XVilka) [reply libr/bin/p/bin_elf.c]: @devnexen you forgot the case when OpenBSD binary can be opened on the other platform.

Comment (XVilka) [reply libr/bin/p/bin_elf.c]: Ah, nevermind, ignore my comment, you mean the case of OpenBSD binary.

**Pull request merged**